core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,165 @@
+  }
+
+  def stop(): Unit = {
+    scheduler.shutdown()
+    scheduler.awaitTermination(10, TimeUnit.SECONDS)
+  }
+
+  // The actual implementation is delegated to strategy
+  private def expireExecutorsInBlackList(
+      failureExecutors: mutable.HashMap[String, FailureStatus]): Unit = synchronized {
+    strategy.expireExecutorsInBlackList(failureExecutors)
+  }
+
+  def updateFailureExecutors(info: TaskInfo, reason: TaskEndReason) : Unit = synchronized {
+    reason match {
+      // If task succeeding, remove related record from failedExecutorMap
+      case Success =>
+        removeFailureExecutors(info.executorId, Some(info.taskId))
+
+      // If task failing, update latest failure time and failedTaskIds
+      case _ =>
+        val executorId = info.executorId
+        failedExecutorMap.get(executorId) match {
+          case Some(failureStatus) =>
+            failureStatus.updatedTime = clock.getTimeMillis()
+            val failedTimes = failureStatus.failedTasks.getOrElse(info.taskId, 0) + 1
+            failureStatus.failedTasks.update(info.taskId, failedTimes)
+          case None =>
+            val failedTasks = mutable.HashMap(info.taskId -> 1)
+            val failureStatus = new FailureStatus(
+              clock.getTimeMillis(),
+              info.host,
+              failedTasks)
+            failedExecutorMap.update(executorId, failureStatus)
+        }
+    }
+  }
+
+  def removeFailureExecutors(
+      executorId: String,
+      taskId: Option[Long]) : Unit = synchronized {
+        taskId match {
+          // If taskId is provided, remove the taskId from failedTasks
+          case Some(id) =>
+            failedExecutorMap.get(executorId).map(fs => {
+              fs.updatedTime = clock.getTimeMillis()
+              fs.failedTasks.remove(id)
+              if(fs.failedTasks.isEmpty){
+                failedExecutorMap.remove(executorId)
+              }
+            })
+          // if taskId is not provided, remove the whole record related to given executorId
+          case None =>
+            failedExecutorMap.remove(executorId)
+        }
+  }
+
+  def executorIsBlacklisted(
+      executorId: String,
+      sched: TaskSchedulerImpl,
+      taskId: Option[Long]) : Boolean = {
+
+    executorBlacklist(sched, taskId).contains(executorId)
+  }
+
+  // The actual implementation is delegated to strategy
+  def executorBlacklist(
+      sched: TaskSchedulerImpl,
+      taskId: Option[Long]): Set[String] = synchronized {
+
+    // If the node is in blacklist, all executors allocated on that node will
+    // also be put into  executor blacklist.
+    // By default it's turned off, user can enable it in sparkConf.
+    val speculationFailedExecutor: Set[String] =
+      if (sparkConf.getBoolean("spark.scheduler.blacklist.speculate", false)) {
+      Set.empty[String]
+    } else {
+      nodeBlacklist(taskId).flatMap(sched.getExecutorsAliveOnHost(_)
+          .getOrElse(Set.empty[String])).toSet
+    }
```

Comment:
actually looking at this in more detail now -- I don't really see the point of this conf option.  To me it seems more natural to just always have the executorBlacklist include the nodeBlacklist.  I imagine that you would always want to exclude all executors on a blacklisted node.  If you really want to make this configurable, it could be left to the strategy to figure out.  (That is, the function just calls `strategy.getExecutorBlacklist`, and most strategies would choose to include all blacklisted nodes in this as well.)

But unless there is a good use case, maybe we should just remove the option completely?